syntax = "proto3";

package metloingest;

service MetloIngest {
    rpc ProcessTrace (ApiTrace) returns (ProcessTraceRes);
}

message KeyVal {
    string name = 1;
    string value = 2;
}

message ApiUrl {
    string host = 1;
    string path = 2;
    repeated KeyVal parameters = 3;
}

message ApiRequest {
    string method = 1;
    ApiUrl url = 2;
    repeated KeyVal headers = 3;
    optional string body = 4;
}

message ApiResponse {
    int32 status = 1;
    repeated KeyVal headers = 2;
    optional string body = 3;
}

message ApiMeta {
    string environment = 1;
    bool incoming = 2;
    string source = 3;
    int32 source_port = 4;
    string destination = 5;
    int32 destination_port = 6;
}

message ApiTrace {
    ApiRequest request = 1;
    ApiResponse response = 2;
    ApiMeta meta = 3;
}

message RepeatedString {
    repeated string rep_string = 1;
}

message SqliRes {
    string data = 1;
    string fingerprint = 2;
}
 
message ProcessTraceRes {
    bool block = 1;
    map<string, string> xss_detected = 2;
    map<string, SqliRes> sqli_detected = 3;
    map<string, RepeatedString> sensitive_data_detected = 4;
    map<string, RepeatedString> data_types = 5;
}